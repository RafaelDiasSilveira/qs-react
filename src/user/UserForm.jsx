import React from 'react';
import PropTypes from 'prop-types';

class UserForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      nome: props.user ? props.user.nome : '',
      apelido: props.user ? props.user.apelido : '',
      telefone: props.user ? props.user.telefone : '',
      senha: props.user ? props.user.senha : '',
      errors: [],
    };
  }

  onNomeChange = (e) => {
    const nome = e.target.value;
    this.setState(() => ({ nome }));
  };

  onApelidoChange = (e) => {
    const apelido = e.target.value;
    this.setState(() => ({ apelido }));
  };

  onTelefoneChange = (e) => {
    const telefone = e.target.value;
    this.setState(() => ({ telefone }));
  }

  onSenhaChange = (e) => {
    const senha = e.target.value;
    this.setState(() => ({ senha }));
  }

  // onSubmit = (e) => {
  //   e.preventDefault();

  //   let validForm = true;
  //   let errors = [];

  //   // TODO: Fazer função para validação e função pra remover espaços duplos
  //   if (!this.state.titulo || this.state.titulo.length >= 20) {
  //     const error = 'Título inválido';
  //     errors = errors.concat(error);
  //   }

  //   if (errors.length > 0) {
  //     this.setState(() => ({ errors }));
  //     validForm = false;
  //   }

  //   if (validForm) {
  //     this.setState(() => ({ errors: [] }));
  //     this.props.onSubmit({
  //       titulo: this.state.titulo,
  //       autor: this.state.autor,
  //       dataPublicacao: this.state.dataPublicacao,
  //       genero: this.state.genero,
  //     });
  //   }
  // };

  render() {
    return (
      <div>

        { this.state.errors.length > 0 &&
          (
            <div className="alert alert-danger" role="alert">
              Erros encontrado:
              {this.state.errors}
            </div>
          )
        }

        <form onSubmit={this.onSubmit}>

          <div className="form-row">
            <div className="col form-group">
              <label>Nome</label>
              <input
                type="text"
                placeholder="Nome"
                value={this.state.nome}
                onChange={this.onNomeChange}
                className="form-control"
              />
            </div>

            <div className="col form-group">
              <label>Apelido</label>
              <input
                type="text"
                placeholder="Apelido"
                value={this.state.apelido}
                onChange={this.onApelidoChange}
                className="form-control"
              />
            </div>
          </div>

          <div className="form-row">
            <div className="col form-group">
              <label>Telefone</label>
              <input
                type="text"
                placeholder="(XX) XXXXX-XXXX"
                className="form-control"
                value={this.state.telefone}
                onChange={this.onTelefoneChange}
              />
            </div>

            <div className="col form-group">
              <label>Senha</label>
              <input
                type="password"
                className="form-control"
                value={this.state.senha}
                onChange={this.onSenhaChange}
              />
            </div>
          </div>

          <button className="btn btn-success btn-lg btn-block">Cadastrar-se</button>

        </form>
      </div>
    );
  }
}

UserForm.propTypes = {
  user: PropTypes.shape({
    nome: PropTypes.string,
    apelido: PropTypes.string,
    telefone: PropTypes.string,
    senha: PropTypes.string,
  }),
  // onSubmit: PropTypes.func,
};

UserForm.defaultProps = {
  user: undefined,
  // onSubmit: () => {},
};

export default UserForm;
