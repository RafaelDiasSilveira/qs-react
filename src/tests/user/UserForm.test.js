import React from 'react';
import { shallow } from "enzyme";

import UserForm from '../../user/UserForm';
import { users } from "../fixtures/fixtures";

let wrapper, onSubmit;

beforeAll(() => {
  onSubmit = jest.fn;
  wrapper = shallow(<UserForm onSubmit={onSubmit} />);

});

describe('should render UserForm correctly', () => {
  test('with data', () => {
    const wrapper = shallow(
      <UserForm
        user={users[0]}
        onSubmit={onSubmit}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  test('without data', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe('should set state correctly on input change', () => {
  test('when setting nome input', () => {
    const value = 'Teste do GG';

    wrapper.find('input').at(0).simulate('change', {
      target: { value }
    });

    expect(wrapper.state('nome')).toBe(value);
  });

  test('when setting apelido input', () => {
    const value = 'Teste do GG';

    wrapper.find('input').at(1).simulate('change', {
      target: { value }
    });

    expect(wrapper.state('apelido')).toBe(value);
  });

  test('when setting telefone input', () => {
    const value = '(61) 98245-1511';

    wrapper.find('input').at(2).simulate('change', {
      target: { value }
    });

    expect(wrapper.state('telefone')).toBe(value);
  });

  test('when setting senha input', () => {
    const value = 'senhaSecreta';

    wrapper.find('input').at(3).simulate('change', {
      target: { value }
    });

    expect(wrapper.state('senha')).toBe(value);
  });

})

// TODO: Teste do submit do form e validação de dados

