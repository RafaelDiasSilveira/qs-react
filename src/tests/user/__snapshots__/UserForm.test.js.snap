// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render UserForm correctly with data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-row"
    >
      <div
        className="col form-group"
      >
        <label>
          Nome
        </label>
        <input
          className="form-control"
          onChange={[Function]}
          placeholder="Nome"
          type="text"
          value="Rafael Dias"
        />
      </div>
      <div
        className="col form-group"
      >
        <label>
          Apelido
        </label>
        <input
          className="form-control"
          onChange={[Function]}
          placeholder="Apelido"
          type="text"
          value="Capotinho"
        />
      </div>
    </div>
    <div
      className="form-row"
    >
      <div
        className="col form-group"
      >
        <label>
          Telefone
        </label>
        <input
          className="form-control"
          onChange={[Function]}
          placeholder="(XX) XXXXX-XXXX"
          type="text"
          value="(61) 98245-1511"
        />
      </div>
      <div
        className="col form-group"
      >
        <label>
          Senha
        </label>
        <input
          className="form-control"
          onChange={[Function]}
          type="password"
          value="senhaSecreta"
        />
      </div>
    </div>
    <button
      className="btn btn-success btn-lg btn-block"
    >
      Cadastrar-se
    </button>
  </form>
</div>
`;

exports[`should render UserForm correctly without data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-row"
    >
      <div
        className="col form-group"
      >
        <label>
          Nome
        </label>
        <input
          className="form-control"
          onChange={[Function]}
          placeholder="Nome"
          type="text"
          value=""
        />
      </div>
      <div
        className="col form-group"
      >
        <label>
          Apelido
        </label>
        <input
          className="form-control"
          onChange={[Function]}
          placeholder="Apelido"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="form-row"
    >
      <div
        className="col form-group"
      >
        <label>
          Telefone
        </label>
        <input
          className="form-control"
          onChange={[Function]}
          placeholder="(XX) XXXXX-XXXX"
          type="text"
          value=""
        />
      </div>
      <div
        className="col form-group"
      >
        <label>
          Senha
        </label>
        <input
          className="form-control"
          onChange={[Function]}
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      className="btn btn-success btn-lg btn-block"
    >
      Cadastrar-se
    </button>
  </form>
</div>
`;
